### PR TITLE
Improve path and branch coverage tracking for afl greybox

### PR DESCRIPTION
**Pull request by dliyanage**

### CODE REVIEW

File: afl-fuzz.c [hunk 1]
```diff
@@ -915,54 +915,50 @@ static inline u8 has_new_bits(u8* virgin_map) {
...
+      u8* cur = (u8*)current;
+
+#ifdef __x86_64__
+    for (u8 j = 0; j < 8; j++) 
+#else
+    for (u8 j = 0; j < 4; j++) 
+#endif 
+      if (edg[j] < 0xff && cur[j]) edg[j]++;
+
+    if (unlikely(*current & *virgin)) {
```
Comment (mboehme) [inline afl-fuzz.c]: I think, the `unlikely` can be removed here.

File: afl-fuzz.c [hunk 2]
```diff
@@ -915,54 +915,50 @@ static inline u8 has_new_bits(u8* virgin_map) {
        that have not been already cleared from the virgin map - since this will
        almost always be the case. */
 
-    if (unlikely(*current) && unlikely(*current & *virgin)) {
+    if (unlikely(*current)) {
+      u8* edg = (u8*)edge;
+      u8* cur = (u8*)current;
+
+#ifdef __x86_64__
+    for (u8 j = 0; j < 8; j++)
```
Comment (mboehme) [inline afl-fuzz.c]: Whitespace nit: You probably need to fix indentation for everything until `edge++;`. Add two spaces.

File: afl-fuzz.c [hunk 3]
```diff
@@ -3163,15 +3159,6 @@ static u8 save_if_interesting(char** argv, void* mem, u32 len, u8 fault) {
...
-  u8* trace_mini = ck_alloc(MAP_SIZE >> 3);	
-  minimize_bits(trace_mini, trace_bits);	
-  u32 cksum_mini = hash32(trace_mini, MAP_SIZE >> 3, HASH_CONST);	
-  ck_free(trace_mini);
-  
-  /* Saturated path increment */	
-  if (path_bits[cksum_mini % MAP_SIZE] < 0xFF) path_bits[cksum_mini % MAP_SIZE] ++;
-
 
   if (fault == crash_mode) {
```
Comment (mboehme) [inline afl-fuzz.c]: After here add
```C
/* Keep track of path hit counts for marked paths */
u32 cksum = hash32(trace_bits, MAP_SIZE, HASH_CONST);
if ((unlikely(path_bits[cksum % MAP_SIZE]) && path_bits[cksum % MAP_SIZE]) < 0xFF)
  path_bits[cksum % MAP_SIZE] ++;  
```
(Sorry, my bad)

File: afl-fuzz.c [hunk 4]
```diff
@@ -3181,7 +3168,10 @@ static u8 save_if_interesting(char** argv, void* mem, u32 len, u8 fault) {
     if (!(hnb = has_new_bits(virgin_bits))) {
       if (crash_mode) total_crashes++;
       return 0;
-    }    
+    } 
+    
+    u32 cksum = hash32(trace_bits, MAP_SIZE, HASH_CONST);	
+    if (path_bits[cksum % MAP_SIZE] < 0xFF) path_bits[cksum % MAP_SIZE] ++;
```
Comment (mboehme) [inline afl-fuzz.c]: Substitute this with
```C
/* Mark path as one to keep track of */
if (!path_bits[cksum % MAP_SIZE]) path_bits[cksum % MAP_SIZE] = 1;
```

**Review by mboehme — COMMENTED** [review]
Thanks @dliyanage! Added a few comments.

**Comment by dliyanage** [comment]
Hi Marcel, 
Made the suggested changes and checked the code locally. It works fine and plot_data seems reasonable. Let me share some results on shorter campaigns (probably 24h).

Thanks.

**Review by mboehme — APPROVED** [review]
Fixed whitespaces and added the edge bug fix that we did for the blackbox variant, as well.

**Comment by dliyanage** [comment]
Thanks Marcel. I am merging this PR since it is approved.

**Pull request merged**